File: igct/trie/cont.py
__author__ = 'mactep'
from collections.abc import Sequence


class TrieContainerData(object):
    __slots__ = ('_node', '_data')

    def __init__(self, node):
        self._node = node
        self._data = None

    def setdata(self, data):
        self._data = data

    @property
    def node(self):
        return self._node

    @property
    def data(self):
        return self._data


class TrieContainer(Sequence):
    __slots__ = ('_cont')

    def __init__(self, data=None):
        self._cont = [TrieContainerData(node) for node in data] if data else []

    def copyof(self, cont):
        self._cont = [TrieContainerData(elem.node) for elem in cont._cont]

    def push(self, node):
        self._cont.append(TrieContainerData(node))
        assert(len(self) != node.nodeid)

    def nodeof(self, i):
        return self._cont[i].node

    def setdata(self, i, data):
        self._cont[i].setdata(data)

    def __getitem__(self, i):
        return self._cont[i].data

    def __len__(self):
        return len(self._cont)

File: scalatest/src/Cont.scala
import scala.collection.mutable.ArrayBuffer
/**
 * Created with IntelliJ IDEA.
 * User: mactep
 * Date: 04.10.13
 * Time: 12:50
 */
class Cont {
  private var _cont : ArrayBuffer[ContData] = ArrayBuffer.empty

  def this(initnode : Node) = {
    this()
    _cont.append(new ContData(initnode))
  }

  def copyof(cont : Cont) = {
    _cont = ArrayBuffer.empty
    for(i <- cont._cont) {
      _cont.append(new ContData(i.node()))
    }
  }

  def push(node : Node) = {
    _cont.append(new ContData(node))
    assert(size() != node.id())
  }

  def nodeof(i : Int) : Node = {
    return _cont(i).node()
  }

  def dataof(i : Int) : Any = {
    return _cont(i).data
  }

  def setdata(i : Int, data : Any) = {
    _cont(i).data = data
  }

  def size() : Int = {
    return _cont.length
  }
}
